grpc/server/src/backend.cpp

```diff
@@ -0,0 +1,104 @@
+#include "backend.h"
+
+#include <grpc/grpc.h>
+#include <grpc++/server.h>
+#include <grpc++/server_builder.h>
+#include <grpc++/server_context.h>
+#include <grpc++/security/server_credentials.h>
+
+#include "action/actionrpc_impl.h"
+#include "core/corerpc_impl.h"
+#include "log.h"
+#include "mission/missionrpc_impl.h"
+#include "telemetry/telemetryrpc_impl.h"
+
+namespace dronecore::backend {
+
+int DroneCoreBackend::run(const int mavlink_listen_port)
+{
+    if (int returnCode = connectToVehicle(mavlink_listen_port) != 0) {
+        return returnCode;
+    }
+
+    grpc::ServerBuilder builder;
+    setupPort(builder);
+
+    CoreServiceImpl core(dc);
+    builder.RegisterService(&core);
+
+    Action action(&dc.device());
+    ActionServiceImpl actionService(action);
+    builder.RegisterService(&actionService);
+
+    Mission mission(&dc.device());
+    MissionServiceImpl missionService(mission);
+    builder.RegisterService(&missionService);
+
+    Telemetry telemetry(&dc.device());
+    TelemetryServiceImpl telemetryService(telemetry);
+    builder.RegisterService(&telemetryService);
+
+    server = builder.BuildAndStart();
+    LogInfo() << "Server started";
+    server->Wait();
+
+    return 0;
+}
+
+int DroneCoreBackend::connectToVehicle(const int port)
+{
+    if (!addUdpConnection(port)) {
+        return 1;
+    }
+
+    logUUIDOnTimeout();
+    waitForDiscovery();
+}
+
+bool DroneCoreBackend::addUdpConnection(const int port)
+{
+    dronecore::DroneCore::ConnectionResult connection_result = dc.add_udp_connection(port);
+
+    if (connection_result != DroneCore::ConnectionResult::SUCCESS) {
+        LogErr() << "Connection failed: " << DroneCore::connection_result_str(connection_result);
+        return false;
+    }
+
+    return true;
+}
+
+void DroneCoreBackend::logUUIDOnTimeout()
+{
+    dc.register_on_timeout([](uint64_t uuid) {
+        LogInfo() << "Device timed out [UUID: " << uuid << "]";
+    });
+}
+
+void DroneCoreBackend::waitForDiscovery()
+{
+    LogInfo() << "Waiting to discover device...";
+    auto discoveryFuture = wrapped_register_on_discover();
+    discoveryFuture.wait();
```

Comment:
The problem here is that `std::promise` in C++11 does not support triggering the callback multiple times. Therefore, this will crash when the vehicle is discovered a second time.

Reply:
That's one of those cases where I wish we had streams :D.

I thought it was not called the second time, as the `device` stays the same. I'll see what I can do here. But it sounds like a mutex and it's more complicated than what I would like to have :).

Reply:
`boost::fibers` would support something like that as well.

Reply:
But since we don't depend on boost yet, I'm happy to try to avoid it unless necessary :D. I'm not sure I want to build it for Android/iOS...